src/EventLogger.h

```diff
@@ -0,0 +1,99 @@
+#include <chrono>
+#include <functional>
+
+namespace KafkaW {
+class ProducerTopic;
+}
+
+namespace FileWriter {
+//-----------------------------------------------------------------------------
+/// @brief      Enum class to map NICOS status code values. Close represent the
+/// successiful closere of a NeXus file, Error a generic, non critical error,
+/// Fail the impossibility to start the job.
+///
+enum class StatusCode : int {
+  Start = 0,
+  Close = 1,
+  Error = 2,
+  Fail = 3,
+};
+//------------------------------------------------------------------------------
+/// @brief      Convert the StatusCode into a string
+///
+/// @param[in]  Code  The StatusCode
+///
+/// @return     An upper case string that NICOS can interpret
+///
+std::string convertStatusCodeToString(StatusCode Code);
+
+//------------------------------------------------------------------------------
+/// @brief      Builds a JSON string that contains an event log and emits the
+/// log via the producer. The event log
+/// is described by a code ("START","CLOSE","FAIL","ERROR"), the service and job
+/// ids, the system timestamp at wich the log has been emitted and a message.
+///
+/// @param[in]  Producer      The producer, for example KafkaW::ProducerTopic
+/// @param[in]  Code          The code
+/// @param[in]  ServiceId     The service identifier
+/// @param[in]  JobId         The job identifier
+/// @param[in]  Message       The message
+///
+/// @tparam     ProducerType  { description }
+///
+template <class ProducerType>
+void logEvent(std::shared_ptr<ProducerType> Producer, StatusCode Code,
+              const std::string &ServiceId, const std::string &JobId,
+              const std::string &Message) {
+
+  if (!Producer) {
+    LOG(Sev::Warning, "Can't produce filewriter event log, invalid Producer");
+    return;
+  }
+
+  std::chrono::system_clock::time_point Now = std::chrono::system_clock::now();
+  nlohmann::json Event;
+  Event["type"] = "filewriter_event";
+  Event["code"] = convertStatusCodeToString(Code);
+  Event["timestamp"] = std::chrono::duration_cast<std::chrono::milliseconds>(
+                           Now.time_since_epoch())
+                           .count();
+  Event["service_id"] = ServiceId;
+  Event["job_id"] = JobId;
+  Event["message"] = Message;
+
+  std::string EventMessage = Event.dump();
+  Producer->produce(reinterpret_cast<unsigned char *>(&EventMessage[0]),
+                    EventMessage.size());
+}
+
+//------------------------------------------------------------------------------
+/// @brief      Class for event logger. An EventLogger obects defines a unique
+/// pair of service and job id.
+///
+class EventLogger {
+public:
+  template <class ProducerType>
+  void create(std::shared_ptr<ProducerType> Producer,
+              const std::string &ServiceId, const std::string &JobId) {
+    using namespace std::placeholders;
+    doLogEvent =
+        std::bind(logEvent<ProducerType>, Producer, _1, ServiceId, JobId, _2);
```

Comment:
Can be written in a more idiomatic way by using some `EventLoggerI` and imlementing a concrete EventLogger for that `ProducerType` and keeping the `Producer` itself as member.

Reply:
As we discuss the class is not necessary. The `Steramer`, which is the one that requires it, return any error to the `StreamMaster`, which has all the knowledge required to report the error.